src/client/dfs/dfs.c

```diff
@@ -846,7 +847,12 @@ open_dir(dfs_t *dfs, daos_handle_t th, daos_handle_t parent_oh, int flags,
 	if (!exists)
 		return ENOENT;
 
-	if (!S_ISDIR(entry.mode))
+	/* Check that the opened object is a directory, this might be because
+	 * it's a FIFO which is expected, or because the type has been changed
+	 * from another client since it was created.
+	 */
+
+	if ((S_ISDIR(dir->mode) && !S_ISDIR(entry.mode)))
```

Comment:
im not sure i understand this change. I don't understand the comment either.
what if dir->mode is not set? like in dfs_mount

Reply:
This function is called for both dirs and fifos, so the check is verifying that if we expect it to be a directory then it is a directory.  It should possibly be checking that dir->mode I S_IFMT = entry.mode | S_IFMT.  Essentially that the directory hasn't been replaced with a entry of a different type since it was opened.

Reply:
well as i mentioned what if we are not setting what we expect?  we only do that in dfs_open.
and to make this clear, maybe add explicit checks here for fifo and dir.

Reply:
Looking at this closer I'm not sure the check is needed.  open_dir is either called in dfs_mount on (/) which will always be a directory, or within a transaction.  There might be a race where one client tries to make a directory, but another client makes a file, but the transaction should catch this so maybe this check could just be removed entirely.

The check was here before anyway, I just extended it to not perform the check if the entry in question is not a directory (and therefore is a fifo) but perhaps the check simply isn't needed.

I don't know what would happen if you called dfs_open(S_IFDIR) but without O_CREATE and a file existed at that name, this check would catch that but that's an odd case that's probably not allowed in POSIX anyway.  DFuse will always pass O_CREAT to dfs_open(), for "open" it uses dfs_dup().

Reply:
well we do want to make sure that the entry is of the correct type?
what if one calls opendir on a file? we can't have this succeed.. not sure how that is not allowed. it can be called directly from the DFS API.

Reply:
This is where we check for calling open_dir() on a file.  I've added S_IFDIR in the call from dfs_mount() and added a seperate check for FIFOs.  open_file() already has a check, and open_symlink doesn't support dfs_open() without O_CREATE.